packages/buildpacks/src/buildpacks.ts

```diff
@@ -0,0 +1,180 @@
+      let body = await response.json()
+      return body.blob_url
+    } catch (err) {
+      if (err.statusCode === 404) {
+        cli.error(`${buildpack} is not in the buildpack registry.`, {exit: 1})
+      } else if (err.statusCode) {
+        cli.error(`${err.statusCode}: ${err.message}`, {exit: 1})
+      } else {
+        cli.error(err.message, {exit: 1})
+      }
+    }
+
+    return ''
+  }
+
+  async findUrl(buildpacks: BuildpackResponse[], buildpack: string): Promise<number> {
+    let mappedUrl = await this.registryNameToUrl(buildpack)
+    return lodashFindIndex(buildpacks, (b: BuildpackResponse) => {
+      return b.buildpack.url === buildpack || b.buildpack.url === mappedUrl
+    })
+  }
+
+  async validateUrlNotSet(buildpacks: BuildpackResponse[], buildpack: string) {
+    if (await this.findUrl(buildpacks, buildpack) !== -1) {
+      cli.error(`The buildpack ${buildpack} is already set on your app.`, {exit: 1})
+    }
+  }
+
+  findIndex(buildpacks: BuildpackResponse[], index?: number) {
+    if (index) {
+      return lodashFindIndex(buildpacks, function (b: BuildpackResponse) {
+        return b.ordinal + 1 === index
+      })
+    } else {
+      return -1
+    }
+  }
+
+  async mutate(app: string, buildpacks: BuildpackResponse[], spliceIndex: number, buildpack: string, command: 'add' | 'set' | 'remove'): Promise<BuildpackResponse[]> {
+    let buildpackUpdates = buildpacks.map(function (b: BuildpackResponse) {
+      return {buildpack: b.buildpack.url}
+    })
+
+    let howmany = (command === 'add') ? 0 : 1
+    let urls = (command === 'remove') ? [] : [{buildpack: await this.registryNameToUrl(buildpack)}]
+
+    let indexes: any[] = [spliceIndex, howmany]
+    let array: any[] = indexes.concat(urls)
+    Array.prototype.splice.apply(buildpackUpdates, array)
+
+    return this.put(app, buildpackUpdates)
+  }
+
+  async put(app: string, buildpackUpdates: {buildpack: string}[]): Promise<BuildpackResponse[]> {
+    let buildpacks = await this.heroku.put(`/apps/${app}/buildpack-installations`, {
+      headers: {Range: ''},
+      body: {updates: buildpackUpdates}
+    })
+
+    return this.mapBuildpackResponse(buildpacks)
+  }
+
+  displayUpdate(app: string, remote: string, buildpacks: BuildpackResponse[], action: 'added' | 'set' | 'removed') {
+    if (buildpacks.length === 1) {
+      cli.log(`Buildpack ${action}. Next release on ${app} will use ${this.registryUrlToName(buildpacks[0].buildpack.url)}.`)
+      cli.log(`Run ${color.magenta(push(remote))} to create a new release using this buildpack.`)
+    } else {
+      cli.log(`Buildpack ${action}. Next release on ${app} will use:`)
+      this.display(buildpacks, '  ')
+      cli.log(`Run ${color.magenta(push(remote))} to create a new release using these buildpacks.`)
+    }
+  }
+
+  registryUrlToName(buildpack: string): string {
+    let match = /^https:\/\/([\w\-]+\/[\w\-]+)\.s3\.amazonaws\.com\/buildpacks\/([\w\-]+\/[\w\-]+).tgz$/.exec(buildpack)
```

Comment:
this will match codon-buildpacks too, correct? We don't want that for non `heroku/` buildpacks, because any other won't be synced to Buildkits. OTH, no one uses Buildkits.